### PR TITLE
clients(devtools): skip page load when running just accessibility

### PR DESCRIPTION
**Pull request by connorjclark**
fixes https://github.com/GoogleChrome/lighthouse/issues/1769

First draft of this adds a `skipPageLoad` option to the pass config. Intended usage is for the DevTools Lighthouse panel to set this option when just the Accessibility category is selected. The config module verifies that `global.isDevtools` is true, and that no required artifact has a `pass` or `afterPass`.

Some other options of varying complexity:

* Further require that there is only one pass if using `skipPageLoad`
* Remove `skipPageLoad` from config, move to `passContext` and set based on `global.isDevtools` and required artifacts (or just check that only `onlyCategories` === `['Accessibility']`)

### CODE REVIEW

**Comment by connorjclark** [comment]
> Is there an urgent need for this and are you OK if this implementation completely changes within a few months? :)
> 
> I ask because the focus of my time for the foreseeable future after 3p experiment work wraps up this week is fraggle rock which is exactly about how to handle these situations in a generic/flexible way.
> 
> There's a lot of complexity in this area that I'm afraid a `skipPageLoad` boolean won't convey to users, but if there's an urgent desire to get the a11y category working in DevTools without a refresh I understand it's a much easier case to handle so may I suggest `__internalSkipPageLoadForDevToolsA11y` or something equally unlikely to be confused for proper support? :D

So how about no option, but a `__internalSkipPageLoadForDevToolsA11y` value on the passContext, set based on `global.isDevtools && onlyCategories === ['Accessibility']`? And remove the config verification entirely.

You could revert it completely if it helps with your larger work.

> I also wouldn't close #1769 just with DevTools-only a11y working.

It can stay open. It doesn't solve the problem posed in the title, but it does seem to solve the problem of every user that commented.

> Did you mean `beforePass` instead of `afterPass` here? Accessibility gatherer has an `afterPass` and that seems to be the one that fits this model.

yeah

**Comment by patrickhulce** [comment]
> So how about no option, but a __internalSkipPageLoadForDevToolsA11y value on the passContext, set based on global.isDevtools && onlyCategories === ['Accessibility']? And remove the config verification entirely.

SGTM 👍

**Comment by brendankenny** [comment]
If this is *only* running accessibility, a simpler and more revertable option might be to not use `GatherRunner` at all, but instead in Runner detect that it's that a devtools a11y run and call a different runner file.

It doesn't need to be generic on gatherers, it doesn't need to change emulation state etc, doesn't need base artifacts, doesn't need load error handling and detection, etc, so splitting parts of `GatherRunner` into load/no-load to pull out a clean no-load execution path is a lot of work compared to basically a try/catch around a `Accessibility.afterPass({driver})` call :)

**Comment by connorjclark** [comment]
> doesn't need base artifacts

I think it will need those otherwise many of the items in the report footer will be blank or errored.

**Comment by connorjclark** [comment]
Confirmed working with DevTools

Just need to modify one thing inside devtools, there's a call to `navigate`: https://source.chromium.org/chromium/chromium/src/+/master:third_party/devtools-frontend/src/front_end/lighthouse/LighthousePanel.js;l=394?q=%22await%20resourceTreeModel.navigate(inspectedURL)%22&ss=chromium

**Comment by connorjclark** [comment]
todo: A chromium webtest is the only way verify this will work.

File: lighthouse-core/gather/gather-runner.js [hunk 1]
```diff
@@ -699,6 +699,57 @@ class GatherRunner {
...
+    const artifacts = {};
+
+    const baseArtifacts = await GatherRunner.initializeBaseArtifacts(options);
+
+    try {
+      await driver.connect();
+
+      await GatherRunner.setupDriver(driver, options);
+
+      for (const passConfig of passConfigs) {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gather-runner.js:720]: seems like we could even drop this loop and just create an explicit pass config that only has the a11y gatherer

File: lighthouse-core/runner.js [hunk 2]
```diff
@@ -212,6 +212,19 @@ class Runner {
       requestedUrl,
       settings: runnerOpts.config.settings,
     };
+
+    const internalSkipPageLoadForDevToolsA11y = Boolean(
+      global.isDevtools &&
```
Comment (patrickhulce) [inline lighthouse-core/runner.js:217]: todo: convert this to an explicit parameter `__ internalSkipPageLoadForDevToolsA11y ` or similar that devtools can use to optionally audit the existing page as-is

**Comment by adamraine** [comment]
@connorjclark can we close this?

**Comment by connorjclark** [comment]
https://github.com/GoogleChrome/lighthouse/issues/1769#issuecomment-999111259

**Pull request closed without merging**